pinot-core/src/test/java/org/apache/pinot/queries/LocalJoinQueriesTest.java

```diff
@@ -0,0 +1,689 @@
+          ColumnDataType.STRING, ColumnDataType.STRING, ColumnDataType.INT
+      });
+      List<Object[]> rows = resultsBlock.getRows();
+      assertEquals(rows.size(), 1980);
+      for (int i = 0; i < 20; i++) {
+        Object[] row = rows.get(i);
+        assertEquals(row[0], "10");
+        assertEquals(row[1], "9");
+        assertEquals(row[2], 7);
+      }
+      for (int i = 20; i < 40; i++) {
+        Object[] row = rows.get(i);
+        assertEquals(row[0], "100");
+        assertEquals(row[1], "99");
+        assertEquals(row[2], 97);
+      }
+    }
+    brokerResponse = getBrokerResponse(getQuery(query));
+    brokerResults = brokerResponse.getResultTable().getRows();
+    assertEquals(brokerResults.size(), 7920);
+    assertEquals(brokerResults.get(0), new Object[]{"9", "10", 7});
+  }
+
+  @Test
+  public void testJoinSelectWithTransform() {
+    /*
+      T2.a T1.a T1.b T2.c
+         1    0    1    5
+         1    0    1    5
+         3    1    2    7
+         3    1    2    7
+         5    2    3    9
+         5    2    3    9
+      ...
+        99   49   50  103
+        99   49   50  103
+      (repeat 10 times)
+     */
+    String query = "SELECT T1.a, T1.b, T2.c FROM T1 JOIN T2 ON CAST(T1.a + T1.b AS INT) = T2.a LIMIT 10000";
+    SelectionResultsBlock resultsBlock1 = (SelectionResultsBlock) getOperator(getQuery(query)).nextBlock();
+    SelectionResultsBlock resultsBlock2 = (SelectionResultsBlock) getOperator(getQueryContext(query)).nextBlock();
+    for (SelectionResultsBlock resultsBlock : Arrays.asList(resultsBlock1, resultsBlock2)) {
+      DataSchema dataSchema = resultsBlock.getDataSchema();
+      assertEquals(dataSchema.getColumnNames(), new String[]{"T1.a", "T1.b", "T2.c"});
+      assertEquals(dataSchema.getColumnDataTypes(), new ColumnDataType[]{
+          ColumnDataType.INT, ColumnDataType.LONG, ColumnDataType.STRING
+      });
+      List<Object[]> rows = resultsBlock.getRows();
+      assertEquals(rows.size(), 1000);
+      for (int i = 0; i < 1000; i++) {
+        int keyValue = ((i / 2) % 50) * 2 + 1;
+        Object[] row = rows.get(i);
+        assertEquals(row[0], (keyValue - 1) / 2);
+        assertEquals(row[1], (long) ((keyValue + 1) / 2));
+        assertEquals(row[2], Integer.toString(keyValue + 4));
+      }
+    }
+    BrokerResponseNative brokerResponse = getBrokerResponse(getQuery(query));
+    List<Object[]> brokerResults = brokerResponse.getResultTable().getRows();
+    assertEquals(brokerResults.size(), 4000);
+    assertEquals(brokerResults.get(0), new Object[]{0, 1L, "5"});
+
+    /*
+      T2.b T1.a T1.b T2.c
+         3    1    2    5
+         3    1    2    5
+         5    2    3    7
+         5    2    3    7
+      ...
+       101   50   51  103
+       101   50   51  103
+      (repeat 10 times)
+     */
+    query = "SELECT T1.a, T1.b + T2.c FROM T1 JOIN T2 ON CAST(T1.a + T1.b AS LONG) = T2.b LIMIT 10000";
```

Comment:
For the scenario where LHS is a function expression in the `JOIN` condition, does it make sense to add a scenario where the function expression consists of both `leftTable` and `rightTable` ?

Something like ....

```
T1 JOIN T2
ON
CAST (T1.a  + T1.b + T2.c AS LONG) = T2.b
```